### PR TITLE
optim: use _get_closed_form_lr in normal step when get_lr not overridden

### LINKED ISSUE #184517: LRScheduler doesn't use _get_closed_form_lr when initalizing scheduler
### 🐛 Describe the bug

When defining a class that inherits from `LRScheduler `, it can't be done by defining only  a `_get_closed_form_lr()` method, as in the initialization `_update_lr()` only uses `get_lr()`. I think it should also use the closed form.

A minimal example that uses a non multiplicative lambda schedule:

```
import torch
from typing import Callable

class BetterLambdaLRScheduler(torch.optim.lr_scheduler.LRScheduler):
    def __init__(self, optimizer: torch.optim.Optimizer, lr_lambda: Callable[[int], float | torch.Tensor], last_epoch: int =-1):
    self.lr_lambda = lr_lambda        
    super().__init__(optimizer, last_epoch=last_epoch)
        

    def _get_closed_form_lr(self) -> list[float | torch.Tensor]:
        return [
            self.lr_lambda(self.last_epoch)
            for _ in self.base_lrs
        ]
```

### Versions

Collecting environment information...
PyTorch version: 2.11.0+cu130
Is debug build: False
CUDA used to build PyTorch: 13.0
ROCM used to build PyTorch: N/A

OS: Ubuntu 22.04.3 LTS (x86_64)
GCC version: (Ubuntu 11.4.0-1ubuntu1~22.04.3) 11.4.0
Clang version: Could not collect
CMake version: version 3.22.1
Libc version: glibc-2.35

Python version: 3.12.11 (main, Jul 23 2025, 00:34:44) [Clang 20.1.4 ] (64-bit runtime)
Python platform: Linux-6.8.0-100-generic-x86_64-with-glibc2.35
Is CUDA available: True
CUDA runtime version: Could not collect
CUDA_MODULE_LOADING set to: 
GPU models and configuration: GPU 0: NVIDIA A100-SXM4-40GB
Nvidia driver version: 590.48.01
cuDNN version: Could not collect
Is XPU available: False
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True
Caching allocator config: N/A

CPU:
Architecture:                            x86_64
CPU op-mode(s):                          32-bit, 64-bit
Address sizes:                           43 bits physical, 48 bits virtual
Byte Order:                              Little Endian
CPU(s):                                  256
On-line CPU(s) list:                     0-255
Vendor ID:                               AuthenticAMD
Model name:                              AMD EPYC 7742 64-Core Processor
CPU family:                              23
Model:                                   49
Thread(s) per core:                      2
Core(s) per socket:                      64
Socket(s):                               2
Stepping:                                0
Frequency boost:                         enabled
CPU max MHz:                             2250.0000
CPU min MHz:                             1500.0000
BogoMIPS:                                4499.52
Flags:                                   fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush mmx fxsr sse sse2 ht syscall nx mmxext fxsr_opt pdpe1gb rdtscp lm constant_tsc rep_good nopl nonstop_tsc cpuid extd_apicid aperfmperf rapl pni pclmulqdq monitor ssse3 fma cx16 sse4_1 sse4_2 x2apic movbe popcnt aes xsave avx f16c rdrand lahf_lm cmp_legacy svm extapic cr8_legacy abm sse4a misalignsse 3dnowprefetch osvw ibs skinit wdt tce topoext perfctr_core perfctr_nb bpext perfctr_llc mwaitx cpb cat_l3 cdp_l3 hw_pstate ssbd mba ibrs ibpb stibp vmmcall fsgsbase bmi1 avx2 smep bmi2 cqm rdt_a rdseed adx smap clflushopt clwb sha_ni xsaveopt xsavec xgetbv1 xsaves cqm_llc cqm_occup_llc cqm_mbm_total cqm_mbm_local clzero irperf xsaveerptr rdpru wbnoinvd amd_ppin arat npt lbrv svm_lock nrip_save tsc_scale vmcb_clean flushbyasid decodeassists pausefilter pfthreshold avic v_vmsave_vmload vgif v_spec_ctrl umip rdpid overflow_recov succor smca sev sev_es ibpb_exit_to_user
Virtualization:                          AMD-V
L1d cache:                               4 MiB (128 instances)
L1i cache:                               4 MiB (128 instances)
L2 cache:                                64 MiB (128 instances)
L3 cache:                                512 MiB (32 instances)
NUMA node(s):                            2
NUMA node0 CPU(s):                       0-63,128-191
NUMA node1 CPU(s):                       64-127,192-255
Vulnerability Gather data sampling:      Not affected
Vulnerability Indirect target selection: Not affected
Vulnerability Itlb multihit:             Not affected
Vulnerability L1tf:                      Not affected
Vulnerability Mds:                       Not affected
Vulnerability Meltdown:                  Not affected
Vulnerability Mmio stale data:           Not affected
Vulnerability Reg file data sampling:    Not affected
Vulnerability Retbleed:                  Mitigation; untrained return thunk; SMT enabled with STIBP protection
Vulnerability Spec rstack overflow:      Mitigation; Safe RET
Vulnerability Spec store bypass:         Mitigation; Speculative Store Bypass disabled via prctl
Vulnerability Spectre v1:                Mitigation; usercopy/swapgs barriers and __user pointer sanitization
Vulnerability Spectre v2:                Mitigation; Retpolines; IBPB conditional; STIBP always-on; RSB filling; PBRSB-eIBRS Not affected; BHI Not affected
Vulnerability Srbds:                     Not affected
Vulnerability Tsa:                       Not affected
Vulnerability Tsx async abort:           Not affected
Vulnerability Vmscape:                   Mitigation; IBPB before exit to userspace

Versions of relevant libraries:
[pip3] Could not collect
[conda] Could not collect

### PR DESCRIPTION
**Pull request by Priyanshu-byte-coder**
## Description

A subclass that implements only `_get_closed_form_lr` (without overriding `get_lr`) fails at construction time with `NotImplementedError`. This is because `__init__` calls `_initial_step()` → `step()` → `_update_lr(epoch=None)`, which always calls `get_lr()`.

The `epoch is not None` path already prefers `_get_closed_form_lr` when available, but the `epoch is None` path does not.

## Fix

In `_update_lr`, when `epoch is None`, also check for `_get_closed_form_lr` if the subclass has not overridden `get_lr`. This mirrors the existing `epoch is not None` logic and preserves all existing behavior (built-in schedulers all override `get_lr`).

## Test

Added `test_closed_form_lr_only_subclass` in `test/optim/test_lrscheduler.py` that defines a minimal scheduler with only `_get_closed_form_lr`, verifies it initializes without error, and checks the LR values are correct.

Fixes #184517.

### CODE REVIEW

**Comment by Priyanshu-byte-coder** [comment]
@pytorch-dev-infra could you approve the CI workflows? Pure Python fix — makes LRScheduler._update_lr use _get_closed_form_lr when the subclass hasn't overridden get_lr, mirroring the existing epoch-is-not-None path. No behavior change for built-in schedulers.

**Comment by Priyanshu-byte-coder** [comment]
Pushed a follow-up commit to fix the `PYFMT` lint failure — wrapped the long boolean condition in `_update_lr` to fit within the 88-char Black limit. The dtensor and ROCm failures appear to be pre-existing infra flakes unrelated to this change.

**Comment by Priyanshu-byte-coder** [comment]
@albanD @malfet CI waiting on workflow approval for 5 days. LRScheduler \_get\_closed\_form\_lr fix, pure Python. Ping for CI approval.

**Comment by janeyx99** [comment]
Thank you very much for sending a PR.
To avoid you spending time preparing a PR for a feature we're not sure we want to add, our usual process is to discuss on the issue before doing the PR.
In particular, make sure that the issue is marked "actionable" before sending a PR!
See https://github.com/pytorch/pytorch/wiki/The-Ultimate-Guide-to-PyTorch-Contributions for details!
